### PR TITLE
[MM-19261] Show a tooltip on public or private channel names that get truncated

### PR DESCRIPTION
**Pull request by abdusabri**
#### Summary
Shows a tooltip when hovering over a channel name that is long enough to get truncated with ellipses (...)

**Screenshots**

image

image

image

#### Ticket Link
  Fixes https://github.com/mattermost/mattermost-server/issues/12662

### CODE REVIEW

**Comment by marianunez** [comment]
@esethna To be consistent, should the GM also only show the tooltip when they get truncated as well? It currently shows it at all times: 

Screen Shot 2019-10-21 at 10 14 03 AM

**Comment by abdusabri** [comment]
@marianunez, thanks for reporting this. Actually, the way I've implemented it should actually handle group messages consistently like other channel types. Here is an example from my local environment where it doesn't show a tooltip.

image

I will try to reproduce the case you've tested on my local environment and see why it behaves differently in that case.

**Comment by marianunez** [comment]
> @marianunez, thanks for reporting this. Actually, the way I've implemented it should handle group messages consistently like other channel types. Here is an example from my local environment where it doesn't show a tooltip.
> 
> image
> 
> I will try to reproduce the case you've tested on my local environment and see why it behaves differently in that case.

@abdusabri Yes, reviewing the code I see that it should behave the same but on the test server it still is showing for the example above. There must be an edge case that is causing that bug.

File: components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.jsx [hunk 1]
```diff
@@ -41,6 +41,22 @@ export default class SidebarChannelButtonOrLink extends React.PureComponent {
...
+
+    state = {
+        showTooltip: false,
+    }
+
+    componentDidMount() {
+        this.onMounted();
+    }
+
+    onMounted = () => {
```
Comment (jespino) [inline components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.jsx]: I don't get why you create an extra function that is only called from `componentDidMount` that is called `onMounted`, shouldn't be it just in the `componentDidMount`?

Comment (abdusabri) [reply components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.jsx]: Thanks for the feedback @jespino!

There is a linting rule that discourages calling `setState` inside `componentDidMount`, it is Ok if it is in a call back function or in another function like `onMounted`. And i thought this approach is better than disabling the rule.

**Comment by esethna** [comment]
@marianunez @abdusabri yes agree for consistency we should only show the tooltips if the channel/GM is truncated

File: components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.test.jsx [hunk 2]
```diff
@@ -156,27 +91,45 @@ describe('component/sidebar/sidebar_channel_button_or_link/SidebarChannelButtonO
...
+
+                expect(wrapper.find(OverlayTrigger)).toHaveLength(0);
+
+                wrapper.instance().setState({showTooltip: true});
+
+                expect(wrapper.find(OverlayTrigger)).toHaveLength(1);
+                const overlayWrapper = wrapper.find(OverlayTrigger).first();
+                expect(overlayWrapper.prop('overlay').type).toEqual(Tooltip);
+                expect(overlayWrapper.prop('overlay').props.children).toEqual(baseProps.displayName);
+                expect(wrapper).toMatchSnapshot();
```
Comment (saturninoabril) [inline components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.test.jsx]: Can we remove saving snapshots here? I find the tests (`expect`) for `OverlayTrigger` and `Tooltip` valuable enough already.

File: e2e/cypress/integration/channel/channel_name_tooltips_spec.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+                // * Verify that the tooltip is displayed
+                cy.get('div.tooltip-inner').should('be.visible');
+
+                // # Move cursor away from channel
+                cy.get(`#sidebarItem_${loggedUser.id}__${user.id}`).should('be.visible').trigger('mouseout');
+            });
+        });
+    });
+});
```
Comment (saturninoabril) [inline e2e/cypress/integration/channel/channel_name_tooltips_spec.js]: 💯 This is really great. Thanks for adding end-to-end tests!

File: components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.test.jsx (hunk 2 above)
Comment (abdusabri) [reply components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.test.jsx]: Thanks for the great feedback @saturninoabril, appreciated!

Sure, I will do so 🙂

**Comment by abdusabri** [comment]
> > @marianunez, thanks for reporting this. Actually, the way I've implemented it should handle group messages consistently like other channel types. Here is an example from my local environment where it doesn't show a tooltip.
> > image
> > I will try to reproduce the case you've tested on my local environment and see why it behaves differently in that case.
> 
> @abdusabri Yes, reviewing the code I see that it should behave the same but on the test server it still is showing for the example above. There must be an edge case that is causing that bug.

@marianunez, I wasn't able to reproduce the issue locally, even when I initiated a group discussion with the 2 users of exactly the same name.

While testing on the test server, I'd closed the group chat and re-opened it again. Now, it doesn't happen on the test server as well. Here is a screenshot.

image

What are your thoughts? How should we proceed with this?

**Comment by abdusabri** [comment]
> > > @marianunez, thanks for reporting this. Actually, the way I've implemented it should handle group messages consistently like other channel types. Here is an example from my local environment where it doesn't show a tooltip.
> > > image
> > > I will try to reproduce the case you've tested on my local environment and see why it behaves differently in that case.
> > 
> > 
> > @abdusabri Yes, reviewing the code I see that it should behave the same but on the test server it still is showing for the example above. There must be an edge case that is causing that bug.
> 
> @marianunez, I wasn't able to reproduce the issue locally, even when I initiated a group discussion with the 2 users of exactly the same name.
> 
> While testing on the test server, I'd closed the group chat and re-opened it again. Now, it doesn't happen on the test server as well. Here is a screenshot.
> 
> image
> 
> What are your thoughts? How should we proceed with this?

@marianunez, after pushing the latest commit, the issue re-appeared on the testing server. I don't know why, but it seems the first time the server is created/updated the issue appears in some cases, initiating new group messages - or closing and opening existing ones - makes it behave normally 😕 (still unable to reproduce locally though) 

image

**Comment by abdusabri** [comment]
@marianunez, the issue is now fixed 🎉 

I will just merge and push, and that should do it for this PR 🙂

File: components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.jsx [hunk 4]
```diff
@@ -41,6 +41,30 @@ export default class SidebarChannelButtonOrLink extends React.PureComponent {
...
+        if (prevProps.displayName !== this.props.displayName) {
+            this.enableToolTipIfNeeded(true);
+        }
+    }
+
+    enableToolTipIfNeeded = (disableIfAlreadyEnabled = false) => {
+        const element = this.displayNameRef.current;
+        if (element && element.offsetWidth < element.scrollWidth) {
+            this.setState({showTooltip: true});
+        } else if (disableIfAlreadyEnabled) {
```
Comment (marianunez) [inline components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.jsx]: Is this flag necessary? Should it be fine if it always turns if the above condition is not met?

Comment (abdusabri) [reply components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.jsx]: @marianunez, good point!

I wanted to avoid potentially unneeded re-renders when the `enableToolTipIfNeeded` is called in 'componentDidMount', however, I'd missed that the component is actually a `PureComponent`. So, I will remove the flag and always set to `false` if the condition is not met
